src/prof_recent.c

```diff
@@ -0,0 +1,328 @@
+#define JEMALLOC_PROF_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/emitter.h"
+
+ssize_t opt_prof_recent_alloc_max = 0;
+malloc_mutex_t prof_recent_alloc_mtx; /* Protects the fields below */
+ssize_t prof_recent_alloc_max;
+static ssize_t prof_recent_alloc_count = 0;
+static prof_recent_t *prof_recent_alloc_head = NULL;
+static prof_recent_t *prof_recent_alloc_tail = NULL;
+
+void
+prof_recent_alloc_prepare(tsd_t *tsd, prof_tctx_t *tctx) {
+	assert(opt_prof && prof_booted);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	/*
+	 * Increment recent_count to hold the tctx so that it won't be gone
+	 * even after tctx->tdata->lock is released.  This acts as a
+	 * "placeholder"; the real recording of the allocation requires a lock
+	 * on prof_recent_alloc_mtx and is done in prof_recent_alloc_slow (when
+	 * tctx->tdata->lock has been released).
+	 */
+	++tctx->recent_count;
+	assert(!prof_tctx_should_destroy(tsd_tsdn(tsd), tctx));
+}
+
+static void
+decrement_recent_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->recent_count > 0);
+	--tctx->recent_count;
+	if (prof_tctx_should_destroy(tsd_tsdn(tsd), tctx)) {
+		prof_tctx_destroy(tsd, tctx);
+	} else {
+		malloc_mutex_unlock(tsd_tsdn(tsd), tctx->tdata->lock);
+	}
+}
```

Comment:
This is not possible, because it's possible that the `tdata` is destroyed in the `prof_tctx_destroy()` call above. This makes the code look dangerous. This piece of code also appears in `prof_alloc_rollback()` and `prof_free_sampled_object()`. I'm planning to extract it out as a wrapper function and have some comment in it to explain, but what would be a good name? `prof_tctx_try_destroy()`?